gem/optimise.py

```diff
@@ -413,3 +480,565 @@ def aggressive_unroll(expression):
+    :param factor_lists: list of factors representing sums of products
+    :return: gem node
+    """
+    if len(factor_lists) < 3:
+        sums = []
+        for fl in factor_lists:
+            sums.append(reduce(Product, fl, one))
+        return reduce(Sum, sums, Zero())
+    else:
+        mid = int(len(factor_lists) / 2)
+        return Sum(sumproduct_2_node(factor_lists[:mid]), sumproduct_2_node(factor_lists[mid:]))
+
+
+def factorise(factor_lists, quad_ind, arg_ind_flat):
+    """
+    recursively pick the most common factor
+    maybe can Memoize this one
+    :param factors: list of list of factors, representing sum of products
+    :return: optimised list of list of factors
+    """
+    if len(factor_lists) <= 1:
+        return factor_lists
+    # count number of common factors
+    counter = OrderedDict()
+    for fl in factor_lists:
+        fl_set = OrderedDict.fromkeys(fl)
+        if len(fl_set) > 1:
+            # at least two terms in product
+            for factor in fl_set:
+                if factor in counter:
+                    counter[factor] += 1
+                else:
+                    counter[factor] = 1
+
+    if not counter:
+        return factor_lists
+    # find which factor to factorise out first
+    mcf_value = [0, 1]  # (num arg indices, count)
+    mcf = None
+    arg_ind_set = set(arg_ind_flat)
+    for factor, count in counter.items():
+        if count > 1:
+            num_arg_ind = len(set(factor.free_indices) & arg_ind_set)
+            if (num_arg_ind > mcf_value[0]) or (num_arg_ind == mcf_value[0] and count > mcf_value[1]):
+                # prioritize factors with more argument indices
+                mcf = factor
+                mcf_value[0] = num_arg_ind
+                mcf_value[1] = count
+
+    if not mcf:
+        # no common factors
+        return factor_lists
+
+    new_list = []
+    rest = []  # new_list = mcf * [rest] + rest
+    # keep original factor list
+    # this might not be economical as most of the times factorise should help,
+    # and the copying slows down the optimisation quite a bit
+    # before_rest = []
+    for fl in factor_lists:
+        if mcf in fl and len(fl) > 1:
+            # at least two terms in the product
+            # before_rest.append(fl)
+            # new_fl = list(fl)
+            # new_fl.remove(mcf)
+            # rest.append(new_fl)
+            fl.remove(mcf)
+            rest.append(fl)
+        else:
+            new_list.append(fl)
+
+    # before_node = reorder(sumproduct_2_node(before_rest), (quad_ind + arg_ind_flat))
+    # before_flop = count_flop(before_node)
+    # after_node = reorder(sumproduct_2_node(factorise(rest, quad_ind, arg_ind_flat)), (quad_ind + arg_ind_flat))
+    # after_flop = count_flop(Product(mcf, after_node))
+    # if after_flop > before_flop:
+    #     # only factorise if reduces flops
+    #     return factor_lists
+    # new_list.append([mcf, after_node])
```

Comment:
If factorisation does not reduce flop, do not factorise.
There seems to be very little gain in doing this, so not checking any more, just following heuristics.